File: ios/Classes/AsaTool.m
#import "AsaTool.h"
#import <AdServices/AdServices.h>
#import <AppTrackingTransparency/AppTrackingTransparency.h>

@implementation AsaTool

+ (nullable NSString *)attributionToken {
    if (@available(iOS 14.3, *)) {
        NSError *error;
        NSString *token = [AAAttribution attributionTokenWithError:&error];
        if (error) {
            NSLog(@"[FlutterAsaAttribution]: Failed to retrieve attribution token: %@", error.localizedDescription);
        }
        return token;
    } else {
        NSLog(@"[FlutterAsaAttribution]: Only support iOS 14.3 and later");
        return nil;
    }
}

+ (void)requestAttributionWithComplete:(void(^)(NSDictionary * _Nullable data, NSError * _Nullable error))complete {
    if (@available(iOS 14.3, *)) {
        NSError *error;
        NSString *token = [self attributionToken];

        if (token.length > 0) {
            [self requestAttributionWithToken:token complete:complete];
        } else {
            if (complete) {
                complete(nil, error ?: [NSError errorWithDomain:@"app" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Failed to retrieve attribution token"}]);
            }
        }
    } else {
        if (complete) {
            complete(nil, [NSError errorWithDomain:@"app" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"ATTracking Not Allowed"}]);
        }
    }
}

+ (void)requestAttributionWithToken:(NSString *)token complete:(void(^)(NSDictionary * _Nullable data, NSError * _Nullable error))complete {
    NSString *url = @"https://api-adservices.apple.com/api/v1/";
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url]];
    request.HTTPMethod = @"POST";
    [request addValue:@"text/plain" forHTTPHeaderField:@"Content-Type"];
    request.HTTPBody = [token dataUsingEncoding:NSUTF8StringEncoding];

    NSURLSessionDataTask *dataTask = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (complete) complete(nil, error);
            });
            return;
        }

        NSError *jsonError;
        NSDictionary *result = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];

        dispatch_async(dispatch_get_main_queue(), ^{
            if (complete) {
                complete(result ?: @{}, jsonError);
            }
        });
    }];

    [dataTask resume];
}

@end

File: ios/flutter_asa_attribution/Sources/flutter_asa_attribution/AsaTool.swift
import AdServices
import Foundation

enum AsaTool {
  static func attributionToken() -> String? {
    guard #available(iOS 14.3, *) else {
      NSLog("[FlutterAsaAttribution]: Only support iOS 14.3 and later")
      return nil
    }

    do {
      return try AAAttribution.attributionToken()
    } catch {
      NSLog("[FlutterAsaAttribution]: Failed to retrieve attribution token: %@", error.localizedDescription)
      return nil
    }
  }

  static func requestAttribution(completion: @escaping ([String: Any]?, Error?) -> Void) {
    guard #available(iOS 14.3, *) else {
      completion(nil, attributionError("Only support iOS 14.3 and later"))
      return
    }

    guard let token = attributionToken(), !token.isEmpty else {
      completion(nil, attributionError("Failed to retrieve attribution token"))
      return
    }

    requestAttribution(token: token, completion: completion)
  }

  private static func requestAttribution(
    token: String,
    completion: @escaping ([String: Any]?, Error?) -> Void
  ) {
    guard let url = URL(string: "https://api-adservices.apple.com/api/v1/") else {
      completion(nil, attributionError("Invalid attribution URL"))
      return
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.addValue("text/plain", forHTTPHeaderField: "Content-Type")
    request.httpBody = token.data(using: .utf8)

    URLSession.shared.dataTask(with: request) { data, _, error in
      if let error = error {
        DispatchQueue.main.async {
          completion(nil, error)
        }
        return
      }

      var result: [String: Any]?
      if let data = data {
        result = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
      }

      DispatchQueue.main.async {
        completion(result ?? [:], nil)
      }
    }.resume()
  }

  private static func attributionError(_ message: String) -> NSError {
    NSError(domain: "app", code: -1, userInfo: [NSLocalizedDescriptionKey: message])
  }
}
